api_core/google/api_core/bidi.py

```diff
@@ -134,6 +137,69 @@ def __iter__(self):
             yield item
 
 
+class _Throttle(object):
+    """A context manager limiting the total entries in a sliding time window.
+
+    If more than ``entry_cap`` attempts are made to enter the context manager
+    instance in the last ``time window`` seconds, the exceeding requests block
+    until enough time elapses.
+
+    The context manager instances are thread-safe and can be shared between
+    multiple threads. If multiple requests are blocked and waiting to enter,
+    the exact order in which they are allowed to proceed is not determined.
+
+    Example::
+
+        max_three_per_second = Throttle(time_window=1, entry_cap=3)
+
+        for i in range(5):
+            with max_three_per_second as time_waited:
+                print("{}: Waited {} seconds to enter".format(i, time_waited))
+
+    Args:
+        time_window (float): the width of the sliding time window in seconds
+        entry_cap (int): the maximum number of entries allowed in the time window
+    """
+
+    def __init__(self, time_window, entry_cap):
+        if time_window <= 0.0:
+            raise ValueError("time_window argument must be positive")
+
+        if entry_cap < 1:
+            raise ValueError("entry_cap argument must be positive")
+
+        self._time_window = datetime.timedelta(seconds=time_window)
+        self._entry_cap = entry_cap
+        self._past_entries = collections.deque(maxlen=entry_cap)  # least recent first
+        self._entry_lock = threading.Lock()
+
+    def __enter__(self):
+        with self._entry_lock:
```

Comment:
There is a lock, but deque should be threadsafe. Was this done to keep things more straightforward? There is reference to the first element and then removing it in a check, which would have to be changed if we moved to a non-locking impl I think?

Reply:
The initial idea was that the lock would serve as a gatekeeper, and if any thread needs to sleep while holding it, other threads trying to enter would also be forced to wait. With a lock, achieving that would be straightforward.

The lock was placed there before implementing the rest of the logic - one could say to avoid accidentally overlooking an edge case withing the logic, yes.
(deque operations themselves are thread-safe, but the rest of the logic surrounding it might not be, thus I went with a conservative approach)

Here's one risky scenario, assume `_past_entries == [1, 2, 3, 4]` and `access_limit == 3`:

```
T   1    2    3    4    5
----|----|----|----|----|--
      |_____________|
                    ↑A
        |____________|
                     ↑B
```

- Thread A enters the manager at T=4+ε, computes its `cutoff_time`, and determines that entry `1` is now irrelevant. 
- Just before left-popping an item, thread B enters at T=4+2ε, determines the same, and removes `1` from the queue. It then figures out that there are already three entries in the window, and goes to sleep.
- Thread A is resumed, and pops the leftmost item, but that's actually entry `2` now! With only `3` and `4` left in the queue, thread proceeds without waiting, because it only saw two items in its window.

Since the queue length and the entries in it are essentially a shared state, having a lock around that avoids tricky scenarios. Besides, the lock is released almost immediately after new entries are again allowed, as the lock-holding thread is put to sleep for just the right amount of time.

Do you concur, or have I overlooked something?

(BTW, double checking concurrency logic is always appreciated, so thanks!)

Reply:
The lock is by far the easier approach here. There would be some complicated bits to make this lock free. I just was just curious if this was considered :)